frontend/javascripts/oxalis/view/right-menu/mapping_info_view.js

```diff
@@ -66,19 +66,78 @@ type State = {
   didRefreshMappingList: boolean,
 };
 
-// This function mirrors convertCellIdToRGB in the fragment shader of the rendering plane
+// See the shader-side implementation of getElementOfPermutation in segmentation.glsl.js
+// for a detailed description.
+function getElementOfPermutation(
+  index: number,
+  sequenceLength: number,
+  primitiveRoot: number,
+): number {
+  const oneBasedIndex = (index % sequenceLength) + 1.0;
+  const isFirstElement = oneBasedIndex === 1.0;
+
+  // The GLSL implementation of pow is 2**(y * log2(x)) in which
+  // intermediate results can suffer from precision loss. The following
+  // code mimics this behavior to get a consistent coloring in GLSL and
+  // JS.
+  const imprecise = x => new Float32Array([x])[0];
+  function glslPow(x, y) {
+    return Math.floor(imprecise(2 ** (y * imprecise(Math.log2(x)))));
+  }
+  const sequenceValue = glslPow(primitiveRoot, oneBasedIndex) % sequenceLength;
+
+  // Only use sequenceLength if the requested element is the first of the sequence
+  // Otherwise, return the actual sequenceValue
+  return isFirstElement ? sequenceLength : sequenceValue;
+}
+
+// Input in [0,1]
+// Output in [0,1] for r, g and b
+function colormapJet(x: number): [number, number, number] {
+  const r = _.clamp(x < 0.89 ? (x - 0.35) / 0.31 : 1.0 - ((x - 0.89) / 0.11) * 0.5, 0, 1);
+  const g = _.clamp(x < 0.64 ? (x - 0.125) * 4.0 : 1.0 - (x - 0.64) / 0.27, 0, 1);
+  const b = _.clamp(x < 0.34 ? 0.5 + (x * 0.5) / 0.11 : 1.0 - (x - 0.34) / 0.31, 0, 1);
+
+  return [r, g, b];
+}
+
+// From: https://stackoverflow.com/a/54070620/896760
+// Input: r,g,b in [0,1], out: h in [0,360) and s,v in [0,1]
+function rgb2hsv(rgb: [number, number, number]): [number, number, number] {
+  const [r, g, b] = rgb;
+  const v = Math.max(r, g, b);
+  const n = v - Math.min(r, g, b);
+
+  // eslint-disable-next-line no-nested-ternary
+  const h = n !== 0 && (v === r ? (g - b) / n : v === g ? 2 + (b - r) / n : 4 + (r - g) / n);
+  // $FlowIgnore
+  return [60 * (h < 0 ? h + 6 : h), v && n / v, v];
+}
```

Comment:
Very good points! I was a bit sloppy there :see_no_evil: I moved the code into the utils.js module. For the two color-related functions, I also chose `utils.js`, since the `color_generator.js` file is not glsl-related (the glsl code uses the `.glsl.js` file extension). I could have created a new file for these two functions, but I think they are also alright in the utils module :)

Reply:
I think you mistook this file with the shader file. I meant to move the none-shader implementation (`mapping_info_view.js`) of the three helper functions to other files.


Nevertheless, also moving the shader implementation is / was a good idea :smile:

Reply:
Ah, I see :) Yeah, I got that bit wrong. I moved the JS code now too, however, I just decided to move them directly to the glsl counter-parts. Since these two implementations should always stay in sync, I thought, it would make sense to keep them in the same place. Do you agree? Since their existence is mostly GLSL-motivated, I think, it's fine to put them to the glsl spot.

Reply:
Sounds reasonable :+1: